cli/asc.js

```diff
@@ -270,115 +274,91 @@ exports.main = function main(argv, options, callback) {
+  let asconfigPath = optionsUtil.resolvePath(opts.config || "asconfig.json", baseDir);
+  let asconfigFile = path.basename(asconfigPath);
+  let asconfigDir = path.dirname(asconfigPath);
+  let asconfig = getAsconfig(asconfigFile, asconfigDir, readFile);
 
   const seenAsconfig = new Set();
-  seenAsconfig.add(path.join(baseDir, args.config));
+  seenAsconfig.add(asconfigPath);
 
+  const target = opts.target || "release";
   while (asconfig) {
-    // merge target first, then merge options, then merge extended asconfigs
-    if (asconfig.targets && asconfig.targets[target]) {
-      args = optionsUtil.merge(exports.options, asconfig.targets[target], args);
-    }
-    if (asconfig.options) {
-      if (asconfig.options.transform) {
-        // ensure that a transform's path is relative to the current config
-        asconfig.options.transform = asconfig.options.transform.map(p => {
-          if (!path.isAbsolute(p)) {
-            if (p.startsWith(".")) {
-              return path.join(asconfigDir, p);
-            }
-            return require.resolve(p);
-          }
-          return p;
-        });
+    // Merge target first
+    if (asconfig.targets) {
+      const targetOptions = asconfig.targets[target];
+      if (targetOptions) {
+        opts = optionsUtil.merge(exports.options, opts, targetOptions, asconfigDir);
       }
-      args = optionsUtil.merge(exports.options, args, asconfig.options);
+    }
+    // Merge general options
+    const generalOptions = asconfig.options;
+    if (generalOptions) {
+      opts = optionsUtil.merge(exports.options, opts, generalOptions, asconfigDir);
     }
 
-    // entries are added to the compilation
+    // Append entries
     if (asconfig.entries) {
-      for (const entry of asconfig.entries) {
-        argv.push(
-          path.isAbsolute(entry)
-            ? entry
-            // the entry is relative to the asconfig directory
-            : path.join(asconfigDir, entry)
-        );
+      for (let entry of asconfig.entries) {
+        argv.push(optionsUtil.resolvePath(entry, asconfigDir));
       }
     }
 
-    // asconfig "extends" another config, merging options of it's parent
+    // Look up extended asconfig and repeat
     if (asconfig.extends) {
-      asconfigDir = path.isAbsolute(asconfig.extends)
-        // absolute extension path means we know the exact directory and location
-        ? path.dirname(asconfig.extends)
-        // relative means we need to calculate a relative asconfigDir
-        : path.join(asconfigDir, path.dirname(asconfig.extends));
-      const fileName = path.basename(asconfig.extends);
-      const filePath = path.join(asconfigDir, fileName);
-      if (seenAsconfig.has(filePath)) {
-        asconfig = null;
-      } else {
-        seenAsconfig.add(filePath);
-        asconfig = getAsconfig(fileName, asconfigDir, readFile);
-      }
+      asconfigPath = optionsUtil.resolvePath(asconfig.extends, asconfigDir);
```

Comment:
Nice!  I was going to mention that we should allow for node_resolution here like ts does.